### PR TITLE
[MOD-1925] modal volumes: commit on exit

### PR DESCRIPTION
**Pull request by thundergolfer**
### Describe your changes

Details in [MOD-1925](https://linear.app/modal-labs/issue/MOD-1925/add-commit-on-exit-functionality-for-volumefs-volumes)

### Backward/forward compatibility checks

Check these boxes or delete any item (or this section) if not relevant for this PR.

- [x] Client+Server: this change is compatible with old servers
- [x] Client forward compatibility: this change ensures client can accept data intended for later versions of itself

### CODE REVIEW

File: modal/_container_entrypoint.py [hunk 1]
```diff
@@ -740,6 +762,12 @@ def main(container_args: api_pb2.ContainerArguments, client: Client):
         else:
             run_with_signal_handler(call_function_async(function_io_manager, imp_fun))
 
+        # Commit on exit to catch uncommitted volume changes and surface background
+        # commit errors.
+        function_io_manager.volume_commit(
```
Comment (gimaker) [inline modal/_container_entrypoint.py:767]: Is it possible that the user code keeps open handles to files on the volume when this code is called?

Say, if a user spawns a thread that keeps an open file handle in write mode. There's nothing preventing that, right? As is, we will not include that file in the commit (unless you explicitly fsync it first)

Comment (gimaker) [reply modal/_container_entrypoint.py:767]: To get around that, we could extend `VolumeCommitRequest` with a flag to get include everything in the snapshot though, including files with active writers.

**Review by gimaker — APPROVED** [review]
Not very familiar with the container entrypoint code but LGTM

File: modal/_container_entrypoint.py (hunk 1 above)
Comment (thundergolfer) [reply modal/_container_entrypoint.py:767]: Yes that is possible. I think implementing this in such a way that no files can be left open and all dirty pages are synced involves the more challenging 'on shutdown' implementation in the worker.

The quickest fix is https://docs.python.org/3/library/os.html#os.sync 

Seems to work on gvisor. These are produced in the fs on `os.sync()` for two open files:

<kbd>
image
</kbd>

Comment (gimaker) [reply modal/_container_entrypoint.py:767]: We get `fsync`s after `os.sync()`? I did not expect that - I don't see that without gVisor. Interesting.

**Pull request merged**